File: src/axm_uc/data/code_system/runtime.py
"""Portable generated-system host. Only the CLI reads files or writes stdout."""
from collections import deque
from copy import deepcopy
import json


def same(a, b):
    if type(a) in (int, float) and type(b) in (int, float):
        return a == b
    if type(a) is not type(b):
        return False
    if isinstance(a, dict):
        return set(a) == set(b) and all(same(a[k], b[k]) for k in a)
    if isinstance(a, list):
        return len(a) == len(b) and all(same(x, y) for x, y in zip(a, b))
    return a == b


def key(value):
    # Normalize integral floats for state deduplication (True remains boolean).
    if type(value) is float and value.is_integer():
        value = int(value)
    if isinstance(value, dict):
        return "{" + ",".join(json.dumps(k, ensure_ascii=False) + ":" + key(value[k]) for k in sorted(value)) + "}"
    if isinstance(value, list):
        return "[" + ",".join(key(v) for v in value) + "]"
    return json.dumps(value, ensure_ascii=False, allow_nan=False)


class System:
    def __init__(self, config, functions):
        self.config = deepcopy(config)
        self.spec = self.config["system"]
        self.functions = dict(functions)
        self.bindings = {row["event"]: row for row in self.spec["bindings"]}
        self.transitions = {(row["from"], row["event"]): row for row in self.spec["machine"]["transitions"]}

    def _violations(self, model):
        return [row["id"] for row in self.spec["invariants"] if self.functions[row["function"]](deepcopy(model)) is not True]

    def _state(self, value):
        if not isinstance(value, dict) or set(value) != {"phase", "model"} or value["phase"] not in self.spec["machine"]["states"]:
            raise ValueError("STATE_INVALID")
        self.functions["ucSystemIdentity"](value["model"])
        if self._violations(value["model"]):
            raise ValueError("STATE_INVARIANT")
        return deepcopy(value)

    def initial(self):
        return self._state({"phase": self.spec["machine"]["initial_state"], "model": self.functions["ucSystemInitial"]()})

    def step(self, state, event):
        current = self._state(state)
        held = lambda status, **extra: {"status": status, "state": deepcopy(current), "effects": [], **extra}
        if not isinstance(event, dict) or set(event) != {"type", "args"} or not isinstance(event.get("type"), str) or not isinstance(event.get("args"), list):
            return held("HOLD_ARGUMENTS")
        binding = self.bindings.get(event["type"])
        if binding is None:
            return held("HOLD_NO_TRANSITION")
        routes = binding.get("routes", {}).get("events", [event["type"]])
        if not any((current["phase"], signal) in self.transitions for signal in routes):
            return held("HOLD_NO_TRANSITION")
        try:
            args = [deepcopy(current["model"]), *deepcopy(event["args"])]
            if "guard" in binding and self.functions[binding["guard"]](*args) is not True:
                return held("HOLD_GUARD")
            proposed = self.functions[binding["reducer"]](*args)
            self.functions["ucSystemIdentity"](proposed)
            signal = self.functions[binding["routes"]["function"]](deepcopy(proposed)) if "routes" in binding else event["type"]
            if signal not in routes or (current["phase"], signal) not in self.transitions:
                return held("HOLD_EXECUTION", error="ROUTE_UNDECLARED")
            violations = self._violations(proposed)
            if violations:
                return held("HOLD_INVARIANT", violations=violations)
            transition = self.transitions[(current["phase"], signal)]
            return {"status": "APPLIED", "state": {"phase": transition["to"], "model": deepcopy(proposed)},
                    "effects": deepcopy(transition["effects"]), "transition": {k: transition[k] for k in ("from", "event", "to")}}
        except (ValueError, TypeError, OverflowError) as exc:
            code = str(exc)
            if code == "ARGUMENT_COUNT" or code == "VALUE_TYPE" or isinstance(exc, TypeError):
                return held("HOLD_ARGUMENTS")
            return held("HOLD_EXECUTION", error=code[:160])

    def replay(self, events, state=None):
        if not isinstance(events, list) or len(events) > 10000:
            raise ValueError("EVENT_LIMIT")
        current = self.initial() if state is None else self._state(state)
        results = []
        for event in events:
            result = self.step(current, event)
            results.append(result)
            current = result["state"]
        return {"state": current, "results": results}

    def checkpoint(self, events):
        replay = self.replay(events)
        return {"schema": "axm.code-system-checkpoint/v0.1", "system_sha256": self.config["system_sha256"],
                "events": deepcopy(events), "state": replay["state"]}

    def restore(self, checkpoint):
        if not isinstance(checkpoint, dict) or set(checkpoint) != {"schema", "system_sha256", "events", "state"}:
            raise ValueError("CHECKPOINT_INVALID")
        if checkpoint["schema"] != "axm.code-system-checkpoint/v0.1" or checkpoint["system_sha256"] != self.config["system_sha256"]:
            raise ValueError("CHECKPOINT_SYSTEM_MISMATCH")
        self._state(checkpoint["state"])
        replay = self.replay(checkpoint["events"])
        if not same(replay["state"], checkpoint["state"]):
            raise ValueError("CHECKPOINT_REPLAY_MISMATCH")
        return deepcopy(replay["state"])

    def explore(self):
        contract = self.spec["exploration"]
        initial = self.initial()
        queue = deque([(initial, [])])
        seen = {key(initial)}
        edges = applied = holds = frontier = 0
        covered = set()
        accepted_events = set()
        phases = {initial["phase"]}

        def report(status, counterexample=None):
            return {"status": status, "states": len(seen), "edges": edges, "applied": applied, "holds": holds,
                    "depth_frontier": frontier, "max_depth": contract["max_depth"], "phases": sorted(phases),
                    "transitions": [json.loads(row) for row in sorted(covered)], "accepted_events": sorted(accepted_events), "counterexample": counterexample}

        while queue:
            state, trace = queue.popleft()
            if len(trace) == contract["max_depth"]:
                frontier += 1
                continue
            for event in contract["events"]:
                if edges >= contract["max_edges"]:
                    return report("BUDGET_EXHAUSTED")
                result = self.step(state, event)
                edges += 1
                if result["status"] not in {"APPLIED", "HOLD_NO_TRANSITION", "HOLD_GUARD"}:
                    return report("COUNTEREXAMPLE", {"events": trace + [deepcopy(event)], "before": state, "result": result})
                if result["status"] != "APPLIED":
                    holds += 1
                    continue
                applied += 1
                accepted_events.add(event["type"])
                covered.add(key(result["transition"]))
                phases.add(result["state"]["phase"])
                state_key = key(result["state"])
                if state_key not in seen:
                    if len(seen) >= contract["max_states"]:
                        return report("BUDGET_EXHAUSTED")
                    seen.add(state_key)
                    queue.append((result["state"], trace + [deepcopy(event)]))
        return report("BOUNDED_COMPLETE")

    def observe(self):
        scenarios, input_unchanged = [], True
        for scenario in self.spec["scenarios"]:
            current, actual, events = self.initial(), [], []
            for step in scenario["steps"]:
                before_state, before_event = deepcopy(current), deepcopy(step["event"])
                result = self.step(current, step["event"])
                input_unchanged &= same(current, before_state) and same(step["event"], before_event)
                actual.append(result)
                events.append(step["event"])
                current = result["state"]
            replay = self.replay(events)
            cut = len(events) // 2
            restored = self.restore(self.checkpoint(events[:cut]))
            resumed = self.replay(events[cut:], restored)
            scenarios.append({"id": scenario["id"], "steps": actual, "replay_equal": same(replay["results"], actual),
                              "recovery_equal": same(resumed["results"], actual[cut:]) and same(self.restore(self.checkpoint(events)), current)})
        return {"system_sha256": self.config["system_sha256"], "scenarios": scenarios,
                "input_unchanged": input_unchanged, "exploration": self.explore()}

    def verify(self):
        observed, issues = self.observe(), []
        coverage, events = set(), set()
        for expected, actual in zip(self.spec["scenarios"], observed["scenarios"]):
            if not actual["replay_equal"] or not actual["recovery_equal"]:
                issues.append({"code": "REPLAY_OR_RECOVERY", "scenario": expected["id"]})
            for index, (step, output) in enumerate(zip(expected["steps"], actual["steps"])):
                if not all(same(output.get(field), value) for field, value in step["expect"].items()):
                    issues.append({"code": "SCENARIO_EXPECTATION", "scenario": expected["id"], "step": index})
                if output["status"] == "APPLIED":
                    coverage.add(key(output["transition"])); events.add(step["event"]["type"])
        exploration = observed["exploration"]
        if exploration["status"] != "BOUNDED_COMPLETE":
            issues.append({"code": "EXPLORATION_" + exploration["status"]})
        coverage.update(key(row) for row in exploration["transitions"])
        events.update(exploration["accepted_events"])
        required = {key({k: row[k] for k in ("from", "event", "to")}) for row in self.spec["machine"]["transitions"]}
        if required - coverage or {row["event"] for row in self.spec["bindings"]} - events:
            issues.append({"code": "TRANSITION_COVERAGE_INCOMPLETE"})
        if not observed["input_unchanged"]:
            issues.append({"code": "INPUT_MUTATED"})
        return {"status": "PASS" if not issues else "HOLD", "issues": issues, "observations": observed}


def load():
    import hashlib
    from pathlib import Path
    folder = Path(__file__).resolve().parent
    lock = json.loads((folder / "source-lock.json").read_text(encoding="utf-8"))
    for name in ("runtime.py", "module.py", "system.json"):
        if hashlib.sha256((folder / name).read_bytes()).hexdigest() != lock.get(name):
            raise ValueError("RUNTIME_SOURCE_CHANGED:" + name)
    # Execute the checked source bytes, avoiding stale/tampered bytecode caches
    # and global module-name collisions between independent generated products.
    namespace = {"__name__": "uc_code_" + lock["module.py"], "__file__": str(folder / "module.py")}
    exec(compile((folder / "module.py").read_bytes(), str(folder / "module.py"), "exec"), namespace)
    config = json.loads((folder / "system.json").read_text(encoding="utf-8"))
    return System(config, namespace["FUNCTIONS"])


if __name__ == "__main__":
    import sys
    from pathlib import Path
    try:
        system = load()
        if len(sys.argv) == 1 or sys.argv[1:] == ["verify"]:
            result = system.verify()
        elif sys.argv[1:] == ["observe"]:
            result = system.observe()
        elif len(sys.argv) == 3 and sys.argv[1] in {"replay", "checkpoint", "restore"}:
            raw = Path(sys.argv[2]).read_bytes()
            if len(raw) > 1048576:
                raise ValueError("REQUEST_BYTES_LIMIT")
            def invalid_constant(value):
                raise ValueError("JSON_NONFINITE:" + value)
            result = getattr(system, sys.argv[1])(json.loads(raw, parse_constant=invalid_constant))
        else:
            raise ValueError("usage: runtime.py [verify | replay events.json | checkpoint events.json | restore checkpoint.json]")
        output = json.dumps(result, sort_keys=True, ensure_ascii=False, allow_nan=False)
        if len(output.encode()) > 8 * 1048576:
            raise ValueError("RESULT_BYTES_LIMIT")
        print(output)
        if result.get("status") == "HOLD":
            sys.exit(1)
    except (ValueError, TypeError, OSError, KeyError) as error:
        print(json.dumps({"status": "HOLD", "error": str(error)[:300]}), file=sys.stderr)
        sys.exit(2)

File: src/axm_uc/data/code_system/runtime.js
'use strict';

// Portable generated-system host. Only load()/CLI read files or write stdout.
const clone = value => JSON.parse(JSON.stringify(value));
const compare = (a, b) => {
  const x = [...a].map(c => c.codePointAt(0)), y = [...b].map(c => c.codePointAt(0));
  for (let i = 0; i < Math.min(x.length, y.length); i++) if (x[i] !== y[i]) return x[i] - y[i];
  return x.length - y.length;
};
const key = value => value === null || typeof value !== 'object' ? JSON.stringify(value) : Array.isArray(value) ?
  '[' + value.map(key).join(',') + ']' : '{' + Object.keys(value).sort(compare).map(k => JSON.stringify(k) + ':' + key(value[k])).join(',') + '}';
const same = (a, b) => key(a) === key(b);
const exact = (value, fields) => value && typeof value === 'object' && !Array.isArray(value) && same(Object.keys(value).sort(), [...fields].sort());

class System {
  constructor(config, functions) {
    this.config = clone(config); this.spec = this.config.system; this.functions = {...functions};
    this.bindings = new Map(this.spec.bindings.map(row => [row.event, row]));
    this.transitions = new Map(this.spec.machine.transitions.map(row => [key([row.from, row.event]), row]));
  }
  _violations(model) { return this.spec.invariants.filter(row => this.functions[row.function](clone(model)) !== true).map(row => row.id); }
  _state(value) {
    if (!exact(value, ['phase', 'model']) || !this.spec.machine.states.includes(value.phase)) throw Error('STATE_INVALID');
    this.functions.ucSystemIdentity(value.model);
    if (this._violations(value.model).length) throw Error('STATE_INVARIANT');
    return clone(value);
  }
  initial() { return this._state({phase: this.spec.machine.initial_state, model: this.functions.ucSystemInitial()}); }
  step(state, event) {
    const current = this._state(state);
    const held = (status, extra = {}) => ({status, state: clone(current), effects: [], ...extra});
    if (!exact(event, ['type', 'args']) || typeof event.type !== 'string' || !Array.isArray(event.args)) return held('HOLD_ARGUMENTS');
    const binding = this.bindings.get(event.type);
    if (!binding) return held('HOLD_NO_TRANSITION');
    const routes = binding.routes ? binding.routes.events : [event.type];
    if (!routes.some(signal => this.transitions.has(key([current.phase, signal])))) return held('HOLD_NO_TRANSITION');
    try {
      // Check arguments before JSON cloning so invalid numbers cannot become null.
      const args = [clone(current.model), ...event.args];
      if (binding.guard && this.functions[binding.guard](...args) !== true) return held('HOLD_GUARD');
      const proposed = this.functions[binding.reducer](...args);
      this.functions.ucSystemIdentity(proposed);
      const signal = binding.routes ? this.functions[binding.routes.function](clone(proposed)) : event.type;
      if (!routes.includes(signal) || !this.transitions.has(key([current.phase, signal]))) return held('HOLD_EXECUTION', {error: 'ROUTE_UNDECLARED'});
      const violations = this._violations(proposed);
      if (violations.length) return held('HOLD_INVARIANT', {violations});
      const transition = this.transitions.get(key([current.phase, signal]));
      return {status: 'APPLIED', state: {phase: transition.to, model: clone(proposed)}, effects: clone(transition.effects),
        transition: {from: transition.from, event: transition.event, to: transition.to}};
    } catch (error) {
      const code = String(error.message);
      return code === 'ARGUMENT_COUNT' || code === 'VALUE_TYPE' || error instanceof TypeError ? held('HOLD_ARGUMENTS') : held('HOLD_EXECUTION', {error: code.slice(0, 160)});
    }
  }
  replay(events, state = null) {
    if (!Array.isArray(events) || events.length > 10000) throw Error('EVENT_LIMIT');
    let current = state === null ? this.initial() : this._state(state);
    const results = [];
    for (const event of events) { const result = this.step(current, event); results.push(result); current = result.state; }
    return {state: current, results};
  }
  checkpoint(events) {
    const replay = this.replay(events);
    return {schema: 'axm.code-system-checkpoint/v0.1', system_sha256: this.config.system_sha256, events: clone(events), state: replay.state};
  }
  restore(checkpoint) {
    if (!exact(checkpoint, ['schema', 'system_sha256', 'events', 'state'])) throw Error('CHECKPOINT_INVALID');
    if (checkpoint.schema !== 'axm.code-system-checkpoint/v0.1' || checkpoint.system_sha256 !== this.config.system_sha256) throw Error('CHECKPOINT_SYSTEM_MISMATCH');
    this._state(checkpoint.state);
    const replay = this.replay(checkpoint.events);
    if (!same(replay.state, checkpoint.state)) throw Error('CHECKPOINT_REPLAY_MISMATCH');
    return clone(replay.state);
  }
  explore() {
    const contract = this.spec.exploration, initial = this.initial(), queue = [[initial, []]], seen = new Set([key(initial)]);
    const covered = new Set(), phases = new Set([initial.phase]), acceptedEvents = new Set();
    let edges = 0, applied = 0, holds = 0, frontier = 0;
    const report = (status, counterexample = null) => ({status, states: seen.size, edges, applied, holds, depth_frontier: frontier,
      max_depth: contract.max_depth, phases: [...phases].sort(compare), transitions: [...covered].sort(compare).map(row => JSON.parse(row)), accepted_events: [...acceptedEvents].sort(compare), counterexample});
    for (let index = 0; index < queue.length; index++) {
      const [state, trace] = queue[index];
      if (trace.length === contract.max_depth) { frontier++; continue; }
      for (const event of contract.events) {
        if (edges >= contract.max_edges) return report('BUDGET_EXHAUSTED');
        const result = this.step(state, event); edges++;
        if (!['APPLIED', 'HOLD_NO_TRANSITION', 'HOLD_GUARD'].includes(result.status)) return report('COUNTEREXAMPLE', {events: [...trace, clone(event)], before: state, result});
        if (result.status !== 'APPLIED') { holds++; continue; }
        applied++; acceptedEvents.add(event.type); covered.add(key(result.transition)); phases.add(result.state.phase);
        const stateKey = key(result.state);
        if (!seen.has(stateKey)) {
          if (seen.size >= contract.max_states) return report('BUDGET_EXHAUSTED');
          seen.add(stateKey); queue.push([result.state, [...trace, clone(event)]]);
        }
      }
    }
    return report('BOUNDED_COMPLETE');
  }
  observe() {
    const scenarios = []; let inputUnchanged = true;
    for (const scenario of this.spec.scenarios) {
      let current = this.initial(); const actual = [], events = [];
      for (const step of scenario.steps) {
        const beforeState = clone(current), beforeEvent = clone(step.event), result = this.step(current, step.event);
        inputUnchanged = inputUnchanged && same(current, beforeState) && same(step.event, beforeEvent);
        actual.push(result); events.push(step.event); current = result.state;
      }
      const replay = this.replay(events), cut = Math.floor(events.length / 2), restored = this.restore(this.checkpoint(events.slice(0, cut))), resumed = this.replay(events.slice(cut), restored);
      scenarios.push({id: scenario.id, steps: actual, replay_equal: same(replay.results, actual),
        recovery_equal: same(resumed.results, actual.slice(cut)) && same(this.restore(this.checkpoint(events)), current)});
    }
    return {system_sha256: this.config.system_sha256, scenarios, input_unchanged: inputUnchanged, exploration: this.explore()};
  }
  verify() {
    const observed = this.observe(), issues = [], coverage = new Set(), events = new Set();
    this.spec.scenarios.forEach((expected, index) => {
      const actual = observed.scenarios[index];
      if (!actual.replay_equal || !actual.recovery_equal) issues.push({code: 'REPLAY_OR_RECOVERY', scenario: expected.id});
      expected.steps.forEach((step, i) => {
        const output = actual.steps[i];
        if (!Object.entries(step.expect).every(([field, value]) => same(output[field], value))) issues.push({code: 'SCENARIO_EXPECTATION', scenario: expected.id, step: i});
        if (output.status === 'APPLIED') { coverage.add(key(output.transition)); events.add(step.event.type); }
      });
    });
    const exploration = observed.exploration;
    if (exploration.status !== 'BOUNDED_COMPLETE') issues.push({code: 'EXPLORATION_' + exploration.status});
    exploration.transitions.forEach(row => coverage.add(key(row))); exploration.accepted_events.forEach(name => events.add(name));
    const missing = this.spec.machine.transitions.some(row => !coverage.has(key({from: row.from, event: row.event, to: row.to})));
    if (missing || this.spec.bindings.some(row => !events.has(row.event))) issues.push({code: 'TRANSITION_COVERAGE_INCOMPLETE'});
    if (!observed.input_unchanged) issues.push({code: 'INPUT_MUTATED'});
    return {status: issues.length ? 'HOLD' : 'PASS', issues, observations: observed};
  }
}

function load() {
  const fs = require('node:fs'), path = require('node:path'), crypto = require('node:crypto');
  const lock = JSON.parse(fs.readFileSync(path.join(__dirname, 'source-lock.json'), 'utf8'));
  for (const name of ['runtime.js', 'module.js', 'system.json']) {
    if (crypto.createHash('sha256').update(fs.readFileSync(path.join(__dirname, name))).digest('hex') !== lock[name]) throw Error('RUNTIME_SOURCE_CHANGED:' + name);
  }
  delete require.cache[require.resolve('./module.js')];
  return new System(JSON.parse(fs.readFileSync(path.join(__dirname, 'system.json'), 'utf8')), require('./module.js'));
}
module.exports = {System, load, same};
if (require.main === module) {
  try {
    const args = process.argv.slice(2), system = load(); let result;
    if (!args.length || same(args, ['verify'])) result = system.verify();
    else if (same(args, ['observe'])) result = system.observe();
    else if (args.length === 2 && ['replay', 'checkpoint', 'restore'].includes(args[0])) {
      const raw = require('node:fs').readFileSync(args[1]);
      if (raw.length > 1048576) throw Error('REQUEST_BYTES_LIMIT');
      result = system[args[0]](JSON.parse(raw));
    } else throw Error('usage: runtime.js [verify | replay events.json | checkpoint events.json | restore checkpoint.json]');
    const output = JSON.stringify(result);
    if (Buffer.byteLength(output) > 8 * 1048576) throw Error('RESULT_BYTES_LIMIT');
    process.stdout.write(output + '\n');
    if (result.status === 'HOLD') process.exitCode = 1;
  } catch (error) { process.stderr.write(JSON.stringify({status: 'HOLD', error: String(error.message).slice(0, 300)}) + '\n'); process.exitCode = 2; }
}
